### PR TITLE
Delineate public vs private API of esm-framework

### PR DESCRIPTION
**Pull request by brandones**
## Requirements
- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.

#### For changes to apps
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).

#### If applicable
- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/master/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary

I know no one asked for this, but I did most of it a couple weeks ago in what seemed like a lull. I'm just wrapping it so it doesn't get completely stale.

This makes the API documentation pretty much usable. If you see a function in the API docs, then it's there for you to use it. All the weird internal stuff is hidden. In fact it's hidden from the imports too, so you can't accidentally use an internal esm-framework API.

I may have gotten some things wrong about what should be internal vs public, especially with regards to the offline API. I would appreciate corrections.

## Other

**https://github.com/openmrs/openmrs-esm-patient-chart/pull/634 must be merged at the same time**

### CODE REVIEW

**Review by ZacButko — APPROVED** [review]
Quite a lot of work. Thanks for cleaning up @brandones

**Review by manuelroemer — COMMENTED** [review]
I think this looks good generally, thanks a lot for the effort of cleaning stuff up!
Regarding the offline API: It's honestly hard to tell whether everything is declared public/internal correctly, at least here on the GitHub web interface. But I guess even if not, we can retroactively fix it. Though generally, everything that's exported should imo be considered part of the public API.

I have one question though: Both in this PR and in [this one](https://github.com/openmrs/openmrs-esm-patient-chart/pull/634) we're now having imports from `@openmrs/esm-framework/src/internal`. Am I assuming right that the goal is to remove all of these "internal imports" in the long run and only leverage the public API in every package that is **not** esm-framework?
If not, I'd suggest to just leave these internal members public. In the .NET ecosystem, API members like these have been lovingly called "pubternal" (i.e. declared as internal, but effectively public since they are accessible anywhere) and I personally think that such pubternal APIs should just be avoided entirely, if possible. They have the disadvantages of a public API (i.e. you are not allowed to make a breaking change exactly _because_ they are used in foreign modules), but that is now hidden away. This could, for example, make a dev who updates a function in esm-framework wrongly assume that he can introduce a breaking change (exactly because the function is declared as internal), but that would of course break other modules unexpectedly.
If it's more about cleaning up and removing some functions from esm-framework, I'd prefer just marking them as deprecated and then gradually removing them when all depending modules have been updated. That's how I'd do it for some esm-offline stuff which I'm unhappy with, at least.

File: docs/main/config.md [hunk 1]
```diff
@@ -9,10 +9,7 @@ please see the the
 
 ## How to make a frontend module configurable
 
-You should use this part of the OpenMRS Frontend Framework to modules configurable.
-
-Start with `npm install --save-dev @openmrs/esm-framework`. This is a runtime
-dependency, so you should also include it in `peerDependencies`.
+You should use the OpenMRS Frontend Framework to modules configurable.
```
[the remark this answers is not in the document]
Comment (brandones) [reply docs/main/config.md:12]:
```suggestion
You should use the OpenMRS Frontend Framework to make modules configurable.
```

Thanks @ZacButko !

**Review by ibacher — APPROVED** [review]
@brandones Thanks for this! I agree with distinction between "core apps" and non-core apps. I don't see a real need to enforce a hard boundary on exports as long as its clear when someone is importing an internal function.

We should probably have an explicitly disclaim that use of functions not marked as public is at the user's own risk and we make no guarantee of such functionality.

**Pull request merged**